proposals/5/README.md

```diff
@@ -0,0 +1,103 @@
+****---
+title: Decomposition of Jenkins X
+linktitle: Decomposition of Jenkins X
+description: Decomposition of Jenkins X
+type: docs
+weight: 50
+---
+
+## 1. Overview
+
+This document outlines the work for decomposing Jenkins X into a series of reusable libraries.
+
+## 1.1 Motivation
+
+At the moment, the source code that comprises the main JX binary is tightly coupled and difficult to work on.  Changes have
+unintended consequences and this makes the codebase fragile.
+
+One of the design goals in Accelerate is to have a **Loosely Coupled Architecture** - in particular: 
+
+_The goal is for your architecture to support the ability of teams to get their work done—from design through to deployment—without requiring high-bandwidth communication between teams._
+
+## 1.2 Background
+
+The main [JX repository](https://github.com/jenkins-x/jx) contains over 250k lines of code, creating Apps / Extensions / Binary Plugins is difficult without adding a dependency back onto the JX repo.
+
+It should be possible to separate the repository out into a series of reusable libraries so only what is needed can be imported.
+
+## 2. Design
+
+There are currently a number of proposed repositories/modules, some of these are currently in flight.
+
+### Existing Modules
+
+* [jx-api](https://github.com/jenkins-x/jx-api): JX CRDs / Installation Requirements
+* [jx-kube-client](https://github.com/jenkins-x/jx-kube-client): A helper module to create a Kubernetes rest config
+* [jx-vault-client](https://github.com/jenkins-x/jx-vault-client): longer term, it may make sense to deprecate this.
+* [go-scm](https://github.com/jenkins-x/go-scm): All 
+
+### Proposed Modules
+
+* jx-kube: It may make sense to refactor some of the higher level utils/kube functions within JX into its own package to make them reusable.
+* Version Streams: Would these make sense to be moved out of JX
+* Gitter?
+* Updatebot (jx create pr)? This would have value outside of the core JX codebase 
+* Storage APIs?
+* Cloud APIs?
+
```

Comment:
Pipeline/config/metapipeline - which is going to be gnarly. Originally `pkg/tekton` was meant to be pretty self-contained, but over time there's been a lot of blurring between that, `pkg/jenkinsfile`, and `pkg/config` (for `ProjectConfig`), so teasing it out isn't going to be trivial. But we definitely want to do that eventually so that Lighthouse doesn't need to depend in `github.com/jenkins-x/jx/v2` and get updated with every release of jx regardless of whether `PipelineActivity` (which is dealt with by `jx-api`) or metapipeline stuff has changed.

Reply:
This is an interesting one, originally i tried to pull over `ProjectConfig` too into `jx-api`, but had issues (down to the jenkinsfile & tekton stuff).  But it probably does make sense to move that "somewhere"

Reply:
Yeah, I think we're going to have to move all of them together, or at least tease things apart better in there.

Reply:
I was working on a pull request as well to at least move/collect the API around ProjectConfig, but it was not so easy. As @abayer is saying you would need to refactor / tease apart things first, before extraction makes sense. Otherwise, you will have issues with the separation of concerns the the extraction would give you.